centml/compiler/server.py

```diff
@@ -1,20 +1,94 @@
+import os
+import pickle
+from http import HTTPStatus
+import logging
 import uvicorn
-from fastapi import FastAPI
-from pydantic import BaseModel
+from fastapi import FastAPI, UploadFile, HTTPException, BackgroundTasks, Response
+from fastapi.responses import FileResponse
+from fastapi.middleware.gzip import GZipMiddleware
+from centml.compiler.server_compilation import hidet_backend_server, storage_path, CompilationStatus, dir_cleanup
+from centml.compiler import config_instance
+
+logger = logging.getLogger(__name__)
 
 app = FastAPI()
+app.add_middleware(GZipMiddleware, minimum_size=1000)
 
-@app.post("/submit/{model_id}")
-async def submit_handler(model_id: int):
-	return {"message": "Compilation job submitted."}
 
 @app.get("/status/{model_id}")
-async def status_handler(model_id: int):
-	return {"message": "Compilation job finished."}
+async def status_handler(model_id: str):
+    if not os.path.isdir(os.path.join(storage_path, model_id)):
+        return {"status": CompilationStatus.NOT_FOUND}
+
+    if not os.path.isfile(os.path.join(storage_path, model_id, "cgraph.zip")):
+        return {"status": CompilationStatus.COMPILING}
+
+    if os.path.isfile(os.path.join(storage_path, model_id, "cgraph.zip")):
+        return {"status": CompilationStatus.DONE}
+
+    # Something is wrong if we get here
+    raise HTTPException(status_code=HTTPStatus.BAD_REQUEST, detail="Status check: invalid status state.")
+
+
+def background_compile(model_id: str, model: UploadFile, inputs: UploadFile):
+    try:
+        tfx_contents = model.file.read()
+        ei_contents = inputs.file.read()
+    except Exception as e:
+        logger.exception(f"Compilation: error reading serialized content. {e}")
+        dir_cleanup(model_id)
+        return
+    finally:
+        model.file.close()
+        inputs.file.close()
+
+    try:
+        tfx_graph = pickle.loads(tfx_contents)
+        example_inputs = pickle.loads(ei_contents)
+    except Exception as e:
+        logger.exception(f"Compilation: error loading pickled content. {e}")
+        dir_cleanup(model_id)
+        return
+
+    try:
+        # This will save the cgraph to {storage_path}/{model_id}/cgraph.zip
+        hidet_backend_server(tfx_graph, example_inputs, model_id)
+    except Exception as e:
+        logger.exception(f"Compilation: error compiling model. {e}")
+        dir_cleanup(model_id)
+
+
+@app.post("/submit/{model_id}")
+async def compile_model_handler(model_id: str, model: UploadFile, inputs: UploadFile, background_task: BackgroundTasks):
+    try:
+        status = (await status_handler(model_id)).get("status")
```

Comment:
I would suggest not to call the handler directly. Move the status check logic to a different function and call that function both from here and from the `status_handler` function.